src/core/ddsc/tests/typebuilder.c

```diff
@@ -0,0 +1,228 @@
+  dds_topic_unpin (t);
+}
+
+static struct ddsi_domaingv *gv_from_topic (dds_entity_t topic)
+{
+  dds_topic *t;
+  dds_return_t ret = dds_topic_pin (topic, &t);
+  CU_ASSERT_EQUAL_FATAL (ret, DDS_RETCODE_OK);
+  struct ddsi_domaingv *gv = &t->m_entity.m_domain->gv;
+  dds_topic_unpin (t);
+  return gv;
+}
+
+static bool ti_to_pairs_equal (dds_sequence_DDS_XTypes_TypeIdentifierTypeObjectPair *a, dds_sequence_DDS_XTypes_TypeIdentifierTypeObjectPair *b)
+{
+  if (a->_length != b->_length)
+    return false;
+  for (uint32_t n = 0; n < a->_length; n++)
+  {
+    struct DDS_XTypes_TypeObject *to_b = NULL;
+    uint32_t m;
+    for (m = 0; !to_b && m < b->_length; m++)
+    {
+      if (!ddsi_typeid_compare_impl (&a->_buffer[n].type_identifier, &b->_buffer[m].type_identifier))
+        to_b = &b->_buffer[m].type_object;
+    }
+    if (!to_b)
+      return false;
+
+    dds_ostream_t to_a_ser = { NULL, 0, 0, CDR_ENC_VERSION_2 };
+    xcdr2_ser (&a->_buffer[n].type_object, &DDS_XTypes_TypeObject_desc, &to_a_ser);
+    dds_ostream_t to_b_ser = { NULL, 0, 0, CDR_ENC_VERSION_2 };
+    xcdr2_ser (to_b, &DDS_XTypes_TypeObject_desc, &to_b_ser);
+
+    if (to_a_ser.m_index != to_b_ser.m_index)
+      return false;
+    if (memcmp (to_a_ser.m_buffer, to_b_ser.m_buffer, to_a_ser.m_index))
+      return false;
+
+    dds_ostream_fini (&to_a_ser);
+    dds_ostream_fini (&to_b_ser);
+  }
+  return true;
+}
+
+static bool ti_pairs_equal (dds_sequence_DDS_XTypes_TypeIdentifierPair *a, dds_sequence_DDS_XTypes_TypeIdentifierPair *b)
+{
+    if (a->_length != b->_length)
+    return false;
+  for (uint32_t n = 0; n < a->_length; n++)
+  {
+    bool found = false;
+    for (uint32_t m = 0; !found && m < b->_length; m++)
+    {
+      if (!ddsi_typeid_compare_impl (&a->_buffer[n].type_identifier1, &b->_buffer[m].type_identifier1))
+      {
+        if (ddsi_typeid_compare_impl (&a->_buffer[n].type_identifier2, &b->_buffer[m].type_identifier2))
+          return false;
+        found = true;
+      }
+    }
+    if (!found)
+      return false;
+  }
+  return true;
+}
+
+static bool tmap_equal (ddsi_typemap_t *a, ddsi_typemap_t *b)
+{
+  return ti_to_pairs_equal (&a->x.identifier_object_pair_minimal, &b->x.identifier_object_pair_minimal)
+      && ti_to_pairs_equal (&a->x.identifier_object_pair_complete, &b->x.identifier_object_pair_complete)
+      && ti_pairs_equal (&a->x.identifier_complete_minimal, &b->x.identifier_complete_minimal);
+}
+
+#define D(n) TypeBuilderTypes_ ## n ## _desc
+CU_TheoryDataPoints (ddsc_typebuilder, topic_desc) = {
+  // CU_DataPoints (const dds_topic_descriptor_t *, &D(t1), &D(t2), &D(t3), &D(t4), &D(t5), &D(t6), &D(t7), &D(t8),
+  //                                                &D(t9), &D(t10), &D(t11), &D(t12), &D(t13), &D(t14), &D(t15), &D(t16),
+  //                                                &D(t17), &D(t18), &D(t19), &D(t20), &D(t21) ),
+  CU_DataPoints (const dds_topic_descriptor_t *, &D(t21))
```

Comment:
Any particular reason for commenting out `t1` .. `t20`? To me it looks like a quick hack to make it easier to look at `t21` in isolation and then forgetting to uncomment the others.

Also, given that you can use any type id, the error case where the type passed to `ddsi_topic_descriptor_from_type` isn't a valid top-level type ought to be tested. The error path looks fine, so it is presumably no more than a trivial test case.

Reply:
Yes, that was just a quick hack to test that specific case, in [2375e80](https://github.com/eclipse-cyclonedds/cyclonedds/pull/1308/commits/2375e807bd6df63e3c4c1ccf9a93b48f118040f3) I've re-enabled all tests. Thanks for reminding me of testing invalid top-level types: the test-case I've added for that revealed some issues in the error path, I've replaced some asserts by return codes.